---
title: "A tour of psychCNVassoc"
author: "Angela Uzelac"
date: "`r format(Sys.time(), '%d %b %Y')`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Introduction_psychCNVassoc}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(psychCNVassoc)
```

## Introduction

`psychCNVassoc` is an R package developed to explore the impact of pathogenic CNVs on psychiatric disorders. This package is developed for BCB410H: Applied Bioinformatics at the University of Toronto, Canada. This document gives a tour of psychCNVassoc. It was written in R Markdown, using the [knitr](https://cran.r-project.org/package=knitr) package for production. See help(package = "psychCNVassoc") for further details.


To download **psychCNVassoc**, use the following commands:

```{r, eval = FALSE}
install.packages("devtools")
library("devtools")
devtools::install_github("angelauzelac10/psychCNVassoc", build_vignettes = TRUE)
library("psychCNVassoc")
```

To list all sample functions available in the package:

```{r, eval = FALSE}
ls("package:psychCNVassoc")
```

To list all sample datasets available in the package:

```{r, eval = FALSE}
data(package = "psychCNVassoc")
```

To run the Shiny app:
Under construction


## Components

![](./psychCNVassoc_introduction_figure.png)

## Sample Usage

To load the CNV regions from sample_CNV_call.rda, provided in the /data folder:

```{r}
cnv_data <- sample_CNV_call
```

To plot the size distribution of provided CNVs:

```{r}
plotCNVsize(cnv_data)
```

To plot the size distribution of provided CNVs for a specific chromosome:

```{r}
plotCNVsize(cnv_data, chromosome_number = "17")
```

To obtain all the genes that are found in the provided CNVs and visualize the distribution of genic vs non-genic CNVs:

```{r}
gene_list <- getCNVgenes(cnv_data, show_piechart = TRUE)
```

To obtain a gene-disease association table from the obtained list of genes encompassed by CNVs:

```{r}
gene_disease_assoc <- getDiseaseAssoc(gene_list)
```

To visualize which psychiatric disorders are associated with the list of genes, create a wordcloud using:

```{r}
plotDiseaseCloud(gene_disease_assoc)
```


## Package References

Uzelac, A. (2023) psychCNVassoc: Determining the Association Between Copy Number
Variants and Psychiatric Disorders.  Unpublished. URL
<https://github.com/angelauzelac10/psychCNVassoc>.


## References

- Wickham H (2016). _ggplot2: Elegant Graphics for Data Analysis_. Springer-Verlag New York. <https://ggplot2.tidyverse.org>

- Wickham H, François R, Henry L, Müller K, Vaughan D (2023). _dplyr: A Grammar of Data Manipulation_. R package version 1.1.3, <https://CRAN.R-project.org/package=dplyr>.

- Lang D, Chien G (2018). _wordcloud2: Create Word Cloud by 'htmlwidget'_. R package version 0.2.1, <https://CRAN.R-project.org/package=wordcloud2>.

- Feinerer I, Hornik K (2023). _tm: Text Mining Package_. R package version 0.7-11, <https://CRAN.R-project.org/package=tm>.

- Durinck S, Spellman PT, Birney E, Huber W (2009). _Mapping identifiers for the integration of genomic datasets with the R/Bioconductor package biomaRt_. Nature Protocols 4, 1184-1191 .

- Gutierrez-Sacristan A, Hernandez-Ferrer C, Gonzalez J, Furlong L (2023). _psygenet2r: psygenet2r - An R package for querying PsyGeNET and to perform comorbidity studies in psychiatric disorders_. doi:10.18129/B9.bioc.psygenet2r <https://doi.org/10.18129/B9.bioc.psygenet2r>, R package version 1.33.5, <https://bioconductor.org/packages/psygenet2r>.

- BiomaRt Bioconductor R package documentation. (2023). Retrieved November 13, 2023, from <http://useast.ensembl.org/info/data/biomart/biomart_r_package.html>

- Rul CVd (2019). _How to Generate Word Clouds in R_. Medium. <https://towardsdatascience.com/create-a-word-cloud-with-r-bde3e7422e8a>

- Gurbich T, Ilinsky V (2020). _ClassifyCNV: A tool for clinical annotation of copy-number variants_. Scientific Reports, 10(1), Article 1. <https://doi.org/10.1038/s41598-020-76425-3>

- Merico D, <joseph.r.lugo@gmail.com> RZpbJL (2023). _cnvGSA: Gene Set Analysis of (Rare) Copy Number Variants_. doi:10.18129/B9.bioc.cnvGSA, R package version 1.46.0, <https://bioconductor.org/packages/cnvGSA>.

- Silva VHd, Ramos M, Groenen M, Crooijmans R, Johansson A, Regitano L, Coutinho L, Zimmer R, Waldron L, Geistlinger L (2020). _CNVRanger: association analysis of CNVs with gene expression and quantitative phenotypes_. Bioinformatics, 36(3), 972-73. <https://doi.org/10.1093/bioinformatics/btz632>.

- OpenAI. (2023, November 7). Code snippet from conversation with ChatGPT.

- Du J (2022). _CNVds: Analyzing Human CNVs Based on Dosage Sensitivity Scores_. Unpublished. URL <https://github.com/jenydu/CNVds>.
  
- Du J (2023). _CNVRS-study_. Unpublished. URL <https://github.com/jenydu/CNVRS-study>.

- Girdher H (2023). _TDM (Term Document Matrix) and DTM (Document Term Matrix)_. Analytics Vidhya. <https://medium.com/analytics-vidhya/tdm-term-document-matrix-and-dtm-document-term-matrix-8b07c58957e2>

- BioRender (2023). Image created by Uzelac, A. Retrieved November 13, 2023, from <https://app.biorender.com/>


------------------------------------------------------------------------

```{r}
sessionInfo()
```


